### PR TITLE
[JENKINS-58602] Hide JENKINS_SECRET from pod manifest

### PR DESCRIPTION
**Pull request by Vlatombe**
[JENKINS-58602](https://issues.jenkins-ci.org/browse/JENKINS-58602)

### CODE REVIEW

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesPipelineTest.java [hunk 1]
```diff
@@ -420,4 +422,20 @@ public void overrideYaml() throws Exception {
...
     }
+
+    @Test
+    @Issue("JENKINS-58602")
+    public void jenkinsSecretHidden() throws Exception {
+        SemaphoreStep.waitForStart("pod/1", b);
+        List<SlaveComputer> kc = Arrays.stream(r.jenkins.getComputers())
+                .filter(SlaveComputer.class::isInstance)
+                .map(SlaveComputer.class::cast)
+                .collect(Collectors.toList());
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesPipelineTest.java]: or you can use `findAny`

File: src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/jenkinsSecretHidden.groovy [hunk 2]
```diff
@@ -0,0 +1,12 @@
+podTemplate(containers: [
+        containerTemplate(name: 'busybox', image: 'busybox', ttyEnabled: true, command: '/bin/cat'),
+]) {
+    node (POD_LABEL) {
+      semaphore 'pod'
+      stage('Run') {
+        container('busybox') {
+          sh "true"
```
Comment (jglick) [inline src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/jenkinsSecretHidden.groovy]:
```suggestion
          sh 'env | sort'
```

(doing this also outside `container`) and I suspect you will find that this fix does not suffice. Try reverting the change to `KubernetesSlave` and adding

```diff
diff --git a/src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/SecretsMasker.java b/src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/SecretsMasker.java
index 87d6b95a..127c002b 100644
--- a/src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/SecretsMasker.java
+++ b/src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/SecretsMasker.java
@@ -139,6 +139,7 @@ public final class SecretsMasker extends TaskListenerDecorator {
             }
             Pod pod = slave.getTemplate().build(slave);
             Set<String> values = new HashSet<>();
+            values.add(c.getJnlpMac());
             LOGGER.finer(() -> "inspecting " + pod);
             for (Container container : pod.getSpec().getContainers()) {
                 Set<String> secretContainerKeys = new TreeSet<>();
@@ -191,7 +192,7 @@ public final class SecretsMasker extends TaskListenerDecorator {
                     // TODO issue a warning if we did not find values for all of secretContainerKeys
                 }
             }
-            return values.isEmpty() ? null : values;
+            return values;
         }
 
     }
```

Comment (Vlatombe) [reply src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/jenkinsSecretHidden.groovy]: Looks good, though I had to keep the change in `KubernetesSlave` as the `TaskListener` there isn't decorated with `SecretMasker` at this point.

**Comment by jglick** [comment]
Some failures in `KubernetesPipelineOverridenNamespaceTest` _[sic]_.

**Comment by Vlatombe** [comment]
Yes, currently looking into them, it looks like some flushing issue...

**Comment by Vlatombe** [comment]
@jglick I narrowed down the failures to something going wrong with the SecretsMasker. If I disable it as follow, the tests pass (but obviously some others would fail). I feel like this is a flushing issue but I can't pinpoint precisely where...

```
diff --git a/src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/SecretsMasker.java b/src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/SecretsMasker.java
index 127c002b..83d90608 100644
--- a/src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/SecretsMasker.java
+++ b/src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/SecretsMasker.java
@@ -64,27 +64,7 @@ public final class SecretsMasker extends TaskListenerDecorator {
 
     @Override
     public OutputStream decorate(OutputStream logger) throws IOException, InterruptedException {
-        // TODO better to pick up a standard API from credentials-binding (more efficient)
-        // https://github.com/jenkinsci/credentials-binding-plugin/pull/59#discussion_r288735761
-        return new LineTransformationOutputStream() {
-            @Override
-            protected void eol(byte[] b, int len) throws IOException {
-                String s = new String(b, 0, len, StandardCharsets.UTF_8);
-                for (String value : values) {
-                    s = s.replace(value, "********");
-                }
-                logger.write(s.getBytes(StandardCharsets.UTF_8));
-            }
-            @Override
-            public void flush() throws IOException {
-                logger.flush();
-            }
-            @Override
-            public void close() throws IOException {
-                super.close();
-                logger.close();
-            }
-        };
+        return logger;
     }
 
     @Extension

```

**Comment by jglick** [comment]
See https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/103. There are some `sh` steps printing text without a final newline. Try changing e.g.

```sh
cat /whatever
```

into

```sh
cat /whatever && echo
```

**Comment by Vlatombe** [comment]
Yep, looks like it does the trick.

**Comment by jglick** [comment]
I am digging into that bug by the way.

**Pull request merged**